csrc/preseg_passes/reorder_sharded_axis.cpp

```diff
@@ -10,142 +10,177 @@
 #include <device_lower/utils.h>
 #include <fusion.h>
 #include <host_ir/lower.h>
+#include <ir/allocation_utils.h>
 #include <ir/base_nodes.h>
 #include <ir/interface_nodes.h>
 #include <ir/utils.h>
 #include <multidevice/utils.h>
 #include <ops/alias.h>
 #include <ops/arith.h>
+#include <scheduler/utils.h>
+#include <transform_replay.h>
 
 namespace nvfuser::preseg_passes {
 
+namespace {
+
+void transform_and_parallelize_like(TensorView* ref, TensorView* tv) {
+  auto old2new = reorderDIDToFront(ref);
+  propagateDIDTransform(
+      ref, {tv}, /*did_pos=*/old2new.size(), PropagateDirection::kForward);
+  shardAllLike(ref, {tv}, {kParallelTypeDIDs.begin(), kParallelTypeDIDs.end()});
+}
+
+void makeCommunicationLayoutCompliant(
```

Comment:
One way to simplify this would be to always create input and output copy, relying on aliasing to no-op redundant copies.
At present, `markAliasesPreparePass` runs before this pass so that won't happen, but it can be simplified in the future.